### ISSUE TITLE
ESM-only packages come with synthesized default in `moduleResolution: bundler`

### ISSUE THREAD

**Issue by Andarist** [microsoft/TypeScript#54752]
# Bug Report

### 🔎 Search Terms

module resolution bundler synthesized default

### 🕗 Version & Regression Information

- This is the behavior in every version I tried

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?target=99&moduleResolution=100#code/PTAEAEFsHsBMFcA2BTASsgztR8AuBLaAOwC5QAjeI2FAJwCgQIYEUzMjkAPXRscXAENaAc2S52GTj3r18kAA7RauUJFiCMuADSgA3qFzQAyrlr4iI0AF9QAM1rRIoAETrNuALR58iT0c8tc0sXAG5QJlwATwVMMgBmADoAJkSABll3LUSjU2CRelyzCwL6AGNiLTUNLWTQAF5QQQB3QXxVeSUVAAo3Gq8fPwCgkpcASnDImLjQJNSM+izcVNhkO0EkXiWVtY3EXByTYstF-tSi-NluLtU9ayA)

### 💻 Code

```ts
// @moduleResolution: bundler
// @module: esnext
// @target: esnext

import mdast, { toString } from "mdast-util-to-string"; // types: 3.2.0

mdast.toString
toString

const mdast2 = await import("mdast-util-to-string"); // types: 3.2.0

mdast2.default
mdast2.default.toString
mdast2.toString

export {}
```

### 🙁 Actual behavior

`default` export is available here and points to the whole module. I don't think this is right, even with `moduleResolution: bundler` but I can be wrong about this since it's really hard to know what kind of assumptions and goals this mode has when it comes to subtle details

### 🙂 Expected behavior

I don't expect `default` to be available here since this ESM-only package **doesn't** have a default export. 

Note that even if this isn't a bug, `.default` available on that dynamically imported binding should behave in the same way in the IDE as the specifier coming from the import statement. Right now the hover info isn't available on it and go to (type?) definition doesn't work on it with cmd+click.

cc @andrewbranch

**Comment by andrewbranch** [comment]
In my view this is part of #54102. The whole schtick is we have no idea that CJS and ESM are even different things outside of `moduleResolution: node16`. So my plan for #54102 is actually two separate things that go together:

- We need a different or expanded trigger for making ESM vs CJS determinations in the first place.
- We need different controls for how that determination affects stuff. Currently it affects both the module resolution algorithm and ESM->CJS default import checking 100% of the time. Those need to be separable, and based on the results we see in https://andrewbranch.github.io/interop-test/, the latter may need to be separately controllable.

**Comment by andrewbranch** [comment]
Also, `moduleResolution` isn’t supposed to control anything except what file a module specifier resolves to, but right now, `moduleResolution` is what triggers the ESM/CJS format determination. A big part of why I did #54567 was to make it so I can switch that without it directly breaking anyone.

**Comment by Andarist** [comment]
There is something broken here when it comes to how `.default`'s symbol is not working properly in the IDE, right?

**Comment by andrewbranch** [comment]
Yeah, I’m sure the default import of a synthesized symbol on a top-level import declaration is special-cased in `getSymbolAtLocation` / go-to-definition.

**Comment by andrewbranch** [comment]
This was backed out of 5.5—hoping to reintroduce a fix in 5.6 via https://github.com/microsoft/TypeScript/pull/58825.
